Public/Src/Cache/ContentStore/Distributed/NuCache/FlushableCache.cs

```diff
@@ -0,0 +1,165 @@
+        }
+
+        /// <nodoc />
+        public void UnsafeClear()
+        {
+            // Order is important here, inverse order could cause deadlock.
+            using (_flushTurnstile.AcquireSemaphore())
+            using (_exchangeLock.AcquireWriteLock())
+            {
+                if (_cache.Count != 0)
+                {
+                    // Nothing guarantees that some number of updates couldn't have happened in between the last flush
+                    // and this reset, because acquisition of the write lock happens after the flush lock. The only way
+                    // to deal with this situation is to force a flush before this assignment and after locks have been
+                    // acquired. However, this isn't required by our code right now; although it is a simple change.
+                    _cache = new ConcurrentBigMap<ShortHash, ContentLocationEntry>();
+                }
+
+                // There should be no flushes ongoing, or we wouldn't have acquired the lock.
+                Contract.Assert(_flushingCache.Count == 0);
+            }
+        }
+
+        /// <nodoc />
+        public void Store(OperationContext context, ShortHash hash, ContentLocationEntry entry)
+        {
+            using (_exchangeLock.AcquireReadLock())
+            {
+                _cache[hash] = entry;
+            }
+        }
+
+        /// <nodoc />
+        public bool TryGetEntry(ShortHash hash, out ContentLocationEntry entry)
+        {
+            using (_exchangeLock.AcquireReadLock())
+            {
+                // The entry could be a tombstone, so we need to make sure the user knows content has actually been
+                // deleted, which is why we check for null.
+                if (_cache.TryGetValue(hash, out entry))
+                {
+                    _database.Counters[ContentLocationDatabaseCounters.TotalNumberOfCacheHit].Increment();
+                    return entry != null;
+                }
+                else if (_flushingCache.TryGetValue(hash, out entry))
+                {
+                    _database.Counters[ContentLocationDatabaseCounters.TotalNumberOfCacheHit].Increment();
+                    return entry != null;
+                }
+            }
+
+            _database.Counters[ContentLocationDatabaseCounters.TotalNumberOfCacheMiss].Increment();
+
+            return false;
+        }
+
+        /// <nodoc />
+        public async Task FlushAsync(OperationContext context)
+        {
+            // This lock is required to ensure no flushes happen concurrently. We may loose updates if that happens.
+            // AcquireAsync is used so as to avoid multiple concurrent tasks just waiting; this way we return the
+            // task to the thread pool in between.
+            using (await _flushTurnstile.AcquireAsync())
+            {
+                PerformFlush(context);
+            }
+        }
+
+        private void PerformFlush(OperationContext context)

```

Comment:
>PerformFlush [](start = 21, length = 12)

I'm tempted to say this should just be inlined into FlushAsync(). It wasn't immediately clear that this is always called inside the flushTurnstile lock

Reply:
Refactored it out just to reduce nesting. I have added a comment.

---
In reply to: [292140919](https://github.com/microsoft/BuildXL/pull/403#discussion_r292140919) [](ancestors = 292140919)